File: ruby/lib/roman_numerals.rb
ROMAN_CHARACTERS = {
  'X' => 10,
  'V' => 5,
  'I' => 1
}

class Integer
  def to_roman
    ''.tap do |roman_str|
      ROMAN_CHARACTERS.reduce(self) do |current_value, (character, divisor)|
        new_chars, remainder = current_value.roman_divmod(character, divisor)
        roman_str << new_chars
        remainder
      end
    end
  end

  def roman_divmod(roman_character, divisor)
    quotient, remainder = self.divmod divisor

    [roman_character * quotient, remainder]
  end
end


class String
  def to_roman
    self.chars.map { |c| ROMAN_CHARACTERS[c] }.reduce(:+)
  end
end

File: javascript/lib/romanNumerals.js
const _ = require('lodash')

const ROMAN_CHARACTERS = {
  'X': 10,
  'V': 5,
  'I': 1
}

const romanDiv = (value, romanChar, divisor) => {
  return _.repeat(romanChar, value / divisor)
}

const fromNumber = num => {
  var romanStr = ''

  _.reduce(ROMAN_CHARACTERS, (currentValue, divisor, romanChar) => {
    romanStr += romanDiv(currentValue, romanChar, divisor)
    return currentValue % divisor
  }, num)

  return romanStr
}

const fromStr = str => _.sum(_.map(str.split(''), chr => ROMAN_CHARACTERS[chr]))

module.exports = {
  fromNumber: fromNumber,
  fromStr: fromStr
}
